x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/CombineDisjunctions.java

```diff
@@ -0,0 +1,148 @@
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.Equals;
+import org.elasticsearch.xpack.esql.expression.predicate.operator.comparison.In;
+
+import java.time.ZoneId;
+import java.util.ArrayList;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import static org.elasticsearch.xpack.esql.core.expression.predicate.Predicates.combineOr;
+import static org.elasticsearch.xpack.esql.core.expression.predicate.Predicates.splitOr;
+import static org.elasticsearch.xpack.esql.type.EsqlDataTypeConverter.ipToString;
+
+/**
+ * Combine disjunctive Equals or In expression on the same field into an In expression.
+ * This rule looks for both simple equalities:
+ * 1. a == 1 OR a == 2 becomes a IN (1, 2)
+ * and combinations of In
+ * 2. a == 1 OR a IN (2) becomes a IN (1, 2)
+ * 3. a IN (1) OR a IN (2) becomes a IN (1, 2)
+ * Combine disjunctive Equals, In or CIDRMatch expressions on the same field into a CIDRMatch expression.
+ * 4. CIDRMatch(a, ip1) OR CIDRMatch(a, ip2) OR a = ip3 or a IN (ip4, ip5) becomes CIDRMatch(a, ip1, ip2, ip3, ip4, ip5)
+ * <p>
+ * This rule does NOT check for type compatibility as that phase has been
+ * already be verified in the analyzer.
+ */
+public final class CombineDisjunctions extends OptimizerRules.OptimizerExpressionRule<Or> {
+    public CombineDisjunctions() {
+        super(OptimizerRules.TransformDirection.UP);
+    }
+
+    protected static In createIn(Expression key, List<Expression> values, ZoneId zoneId) {
+        return new In(key.source(), key, values);
+    }
+
+    protected static Equals createEquals(Expression k, Set<Expression> v, ZoneId finalZoneId) {
+        return new Equals(k.source(), k, v.iterator().next(), finalZoneId);
+    }
+
+    protected static CIDRMatch createCIDRMatch(Expression k, List<Expression> v) {
+        return new CIDRMatch(k.source(), k, v);
+    }
+
+    @Override
+    public Expression rule(Or or) {
+        Expression e = or;
+        // look only at equals, In and CIDRMatch
+        List<Expression> exps = splitOr(e);
+
+        Map<Expression, Set<Expression>> ins = new LinkedHashMap<>();
+        Map<Expression, Set<Expression>> cidrs = new LinkedHashMap<>();
+        Map<Expression, Set<Expression>> ips = new LinkedHashMap<>();
+        ZoneId zoneId = null;
+        List<Expression> ors = new LinkedList<>();
+        boolean changed = false;
+        for (Expression exp : exps) {
+            if (exp instanceof Equals eq) {
+                // consider only equals against foldables
+                if (eq.right().foldable()) {
+                    ins.computeIfAbsent(eq.left(), k -> new LinkedHashSet<>()).add(eq.right());
+                    if (eq.left().dataType() == DataType.IP) {
+                        Object value = eq.right().fold();
+                        if (value instanceof BytesRef bytesRef) {
```

Comment:
Here you could add a comment mentioning why this should always be a BytesRef here.

Reply:
Yeah, it looks confusing, I'll add some comments here for Equals and In.